src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionProvider.cs

```diff
@@ -78,79 +82,122 @@ private async Task AddCompletionItemsAsync(CompletionContext completionContext,
                     {
-                        typeImportCompletionService.GetTopLevelTypesFromPEReference(
-                            project.Solution,
-                            compilation,
-                            peReference,
-                            GetHandler(compilation.Assembly, assembly),
-                            cancellationToken);
+                        var assemblyProject = project.Solution.GetProject(referencedAssemblySymbol, cancellationToken);
+                        if (assemblyProject != null && assemblyProject.SupportsCompilation)
+                        {
+                            return await typeImportCompletionService.GetTopLevelTypesAsync(assemblyProject,
+                                                                                        syntaxContext,
+                                                                                        GetHandler(compilation.Assembly, referencedAssemblySymbol),
+                                                                                        delayItemHandling: true,
+                                                                                        cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (metadataReference is PortableExecutableReference peReference)
+                        {
+                            return Task.Run<Task>(() => typeImportCompletionService.GetTopLevelTypesFromPEReference(
+                                project.Solution,
+                                compilation,
+                                peReference,
+                                syntaxContext,
+                                GetHandler(compilation.Assembly, referencedAssemblySymbol),
+                                delayItemHandling: true,
+                                cancellationToken), cancellationToken);
+                        }
                     }
+
+                    return Task.CompletedTask;
+                }));
+
+            // We want to timebox the operation that might need to traverse all the type symbols and populate the cache, 
+            // the idea is not to block completion for too long (likely to happen the first time import completion is triggered).
+            // The trade-off is we might not provide unimported types until the cache is warmed up.
+            if (await Task.WhenAny(combinedTask, Task.Delay(TimeoutInMilliseconds, cancellationToken)).ConfigureAwait(false) == combinedTask)
+            {
+                // No timeout. We now have all completion items ready. 
+                var delayedItemHandlingTasks = await combinedTask.ConfigureAwait(false);
+
+                // Tasks for handling items need to run sequentially,
+                // since CompletionContext.AddItem isn't thread safe.
+                foreach (var delayedTask in delayedItemHandlingTasks)
+                {
+                    delayedTask.Start();
```

Comment:
this feels super strange and confusing.  you're 'starting' a task?

Reply:
See https://github.com/dotnet/roslyn/pull/36936#discussion_r299711382